### PR TITLE
Patch from the CMS

### PR DESCRIPTION
**Pull request by wilsonge**

### CODE REVIEW

**Comment by eddieajau** [comment]
George, would mind looking if there is any unit test covering this area of the code? It would be great to add an assertion to cover this case.

**Comment by wilsonge** [comment]
This function (currently) isn't tested on the CMS or framework. I can try and make one for both framework and CMS if you want?

**Comment by eddieajau** [comment]
If you have the time to look at it, that would be great. If you need a hand let me know.

**Comment by wilsonge** [comment]
OK how does this look? I'm not a regex expert - so I'm not 100% sure how to test each bit step by step. But I think this technically gives 100% code coverage for the method. And it can always be built on :)

**Comment by eddieajau** [comment]
Yeah that's pretty close. There's really only two cases we need to test: a trailing dot and an extension plus the regex. So I think you could test that:

`joomla!_config -2.txt` -> `joomla_config -2.txt`
`joomla.` -> `joomla`

The regex allows a scalar (number or letter), a dot, a dash, an underscore and a space.

How does that sound?

**Comment by eddieajau** [comment]
Oh, hang on. You need to move the `rtrim` after the preg_replace because `joomla.!` should filter to `joomla`.

**Comment by wilsonge** [comment]
1. Again not an expert put doesn't that make `joomla!.` fail? if you change the order around? Or does that mean the `rtrim` needs to be before and after the `preg_replace`?
2. Is a space in a file name safe!?

**Comment by eddieajau** [comment]
I think `joomla!.` should filter to `joomla` but you can explicitly test that case too (doesn't hurt). As for the order, it probably should be something like this:

```
$regex = array('#(\.){2,}#', '#[^A-Za-z0-9\.\_\- ]#', '#^\.#');

return rtrim(preg_replace($regex, '', $file), '.');
```

That also begs the question if `.joomla` is valid (I think not). So maybe it should be `trim` rather than `rtrim`?

**Comment by wilsonge** [comment]
.gitignore should be valid imo - whilst an unlikely case there are some uses of files like that

**Comment by eddieajau** [comment]
Good point.

**Comment by wilsonge** [comment]
OK Updated the patch and tests. This just leaves whether having a space in a file name is 'safe'. I have to say when used in the context of uploading files etc. Which this conceivable could - then I don't think that is safe.

But my regex skills are definitely not good enough to fix it to forbid that if you agree!!

**Comment by eddieajau** [comment]
I think spaces are safe at a framework level. The application, like the CMS, can always apply more rules if it wants.

File: src/Joomla/Filesystem/File.php [hunk 1]
```diff
@@ -63,7 +63,10 @@ public static function makeSafe($file)
 	{
 		$regex = array('#(\.){2,}#', '#[^A-Za-z0-9\.\_\- ]#', '#^\.#');
 
-		return preg_replace($regex, '', $file);
+		preg_replace($regex, '', $file);
```
Comment (eddieajau) [inline src/Joomla/Filesystem/File.php]: This line needs to be `$file = preg_replace($regex, '', $file);`.

**Comment by wilsonge** [comment]
God you can tell it was nearly 3am -.-

**Comment by wilsonge** [comment]
Based on what you said about applying more rules I decided to make the regex a param (default is still what it is now) so it can be overridden at the application level if desired to remove spaces or tighten up any of the files above.

For some reason in .gitignore etc. the thing is still failing anyway - will look at that later when I'm back from work

**Comment by eddieajau** [comment]
Hi George. I don't think it's a good idea to put the regex in argument list.

**Comment by wilsonge** [comment]
How about an allowSpaces parameter which them performs an additional regex
to replace spaces with underscores. I really think we should be
incorporating something like that in these checks. To me the majority of
files using this method will be checking suitability of files to be placed
on the server so I strongly feel this should be incorporated somewhere in
the method - even if that allowSpaces parameter is true by default

**Comment by wilsonge** [comment]
Right updated this again with the allowSpaces param with a unit test for it

**Comment by wilsonge** [comment]
OK travis likes this - I like this better - although I don't understand why the .gitignore test is failing. What's stripping the first fullstop :/ Would the regex do that?

**Comment by eddieajau** [comment]
Hi George. I don't think you need this because spaces are actually valid in a file name. So it's appropriate at this low level to allow them. If you just revert that code we should be able to get this merged. That fix with the `rtrim`, however, needs to be pushed back up to the CMS.

**Comment by wilsonge** [comment]
Agree'd with the <code>rtrim</code> stuff - already contacted Mark Dexter about it - but will pull it as well with these tests.

I still disagree with the spaces though. In my opinion - even in the framework - the majority of cases of using makeSafe will be in a file upload capacity to a server - in which case spaces are generally frowned upon. I've turned this param to false as a compromise (i.e. allowing spaces) and in my opinion for one line - it's a useful case the framework should support

Also any ideas about that final test that's failing with .gitignore? Struggling to work it out.

**Comment by eddieajau** [comment]
I understand your argument, I just don't see the space as a big deal anymore - operating systems don't care anymore. However, if one of the other guys wants to support it, I'm ok with it.

Can you paste the test error?

**Comment by wilsonge** [comment]
```
1) JFileTest::testMakeSafe with data set #5 ('.gitignore', true, '.gitignore', 'Files starting with a fullstop should be allowed')
Files starting with a fullstop should be allowed
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'gitignore'
+'.gitignore'
```

And OK i'm happy to go with general opinion :) It's not a major feature thing - so I don't wanna cause a huge stink over it

**Comment by eddieajau** [comment]
George, what about pass a string of characters to strip/exclude rather than just the space. That might be agood compromise?

`public static function makeSafe($file, $stripChars = '')`

If you want to strip spaces or the letter 'A' or whatever, you can do so.

**Comment by wilsonge** [comment]
Sounds like a good plan to me :) I'll work on it!

**Comment by wilsonge** [comment]
Hi Andrew, Successfully implemented with a test (again for spaces) - but this is all generic so more tests can be added later. .gitignore still failing - I can only think it's the regex - so can you check again please?

File: src/Joomla/Filesystem/File.php [hunk 2]
```diff
@@ -51,19 +51,35 @@ public static function stripExt($file)
...
+	 * @param   array   $stripChars  Array with key of string to remove and value with string to replace
 	 *
 	 * @return  string  The sanitised string
 	 *
 	 * @since   1.0
 	 */
-	public static function makeSafe($file)
+	public static function makeSafe($file, $stripChars = array())
 	{
 		$regex = array('#(\.){2,}#', '#[^A-Za-z0-9\.\_\- ]#', '#^\.#');
```
Comment (eddieajau) [inline src/Joomla/Filesystem/File.php]: I think you'll find the `'#^\.#'` is causing the problem with `.gitignore`. Only trouble is removing that creates a backward incompatibility (technically). I'm also thinking that we might want to exclude dot-files and we can't do that with an array of strip-chars.

Sooooo, we could turn $stripChars into an array of regexes, something like:

```
public static function makeSafe($file, array $replacements = array('#^\.#'))
{
     $regex = array_merge(array('#(\.){2,}#', '#[^A-Za-z0-9\.\_\- ]#'), $replacements);
```

That will maintain default behaviour but also allow you to write whatever freaky regexes to make the file name comply with your rules. In the case of `.gitignore`, you'd test `$file->makeSafe('.gitignore', array())`.

Or, we say this is beta and we can change what we want and make `$replacements = array()`, adding a suggestion for excluding hidden dot-files in the docblock. I'm not sure what the best way is (though, the former will be required if you push it back to the CMS).

**Comment by wilsonge** [comment]
OK I've implemented this - however this now has the rule that the spaces are just removed rather than replaced with underscores (as we're not using preg_replace obviously) - is this acceptable? Personally I'm not as keen - but it does make more sense

**Comment by wilsonge** [comment]
Hi Andrew - quick bump on this - know your busy in the apps debate - but I'd like to push this back into the CMS for 3.2 once we've made a decision :)

**Comment by dongilbert** [comment]
I like the change overall as well. @eddieajau is busy with Joomla Day Sydney this weekend, so I don't expect to hear back from him, but when we do, and if he's good with it, I'll get it merged.

**Comment by eddieajau** [comment]
Thanks George.

**Comment by wilsonge** [comment]
Thankyou for all your help (y)

**Pull request merged**